Merge code: use memory watch instead of file watch

votuanphong1983 opened the pull request:
Refer ticket: [https://freedom.myjetbrains.com/youtrack/issue/VN-133](https://freedom.myjetbrains.com/youtrack/issue/VN-133)

This is a fork version of electron-settings with some below changes

- Use  lodash to simplify the code. Make code easier to understand
- Use memory watcher instead of file watcher. Make watcher more efficient, we do not have to read file every time we want to get a property
- Support unwatch api that allow user to unregiester his watcher

**Please note:** because this library watches on memory instead of file. Please not that the change event is not fired when you change value directly on setting file

votuanphong1983 committed:
Merge code: use memory watch instead of file watch

votuanphong1983 committed:
fix typo

pandafox commented on lib/settings-observer.js:
```diff
@@ -91,7 +74,7 @@ class SettingsObserver {
    */
   dispose() {
-    this._settings.removeListener('change', this._handleChange);
+    this._settings._watchHandlers = _.reject(this._settings._watchHandlers, n => n === this);
```
this._settings._watchHandlers = []

votuanphong1983 replied:
nope...this method just dispose/unwatch only 1 observer. i do not think we clear all our watcher list

pandafox replied:
If that's the case, the setup settings seem very odd since we don't really allow adding multiple handlers to the observer anyways? This should just be one handler per observer?

votuanphong1983 replied:
1 observer stores. 
  + keyPath
  + 1 handler
  + current value (we need this value to compare)
  ... others..

I applied this solution from the beginning.  E.g:
 watch('object.propA', handler1) => observer1
 watch('object.propA', handler2) => observer2
 watch('object.propB', handler3) => observer3
 watch('object.propB.propB1, handler4) => observer4
...

pandafox replied:
Ah okay, this is removing from the list of handlers (which is a bit odd). The way this is accessed is a bit odd since _settings._watchHandlerrs are both suppose to be the private variable of the object, but they are not private of the observer, but private to the settings?

votuanphong1983 replied:
"but they are not private of the observer, but private to the settings?" => yeah...exactly => the private property is confusing..

pandafox replied:
In this case, the observer should call the settings or its container to remove it?

(ex: unwatch(observer)?

votuanphong1983 replied:
actually i think we can remove observer.dispose method. Actually we do not want it i think... because in setting, we can setting.unwatch(observer).

I would like to keep observer.dispose because the testcases that relate to watch. In testcase, they call dispose after finished a case => Finally i think we should
 - remove observer dispose
 - update testcases (do not call observer.dispose(). call setting.unwatch instead)

pandafox commented on lib/settings.js:
```diff
@@ -333,34 +186,14 @@ class Settings extends EventEmitter {
-        Helpers.deleteValueAtKeyPath(obj, keyPath);
-        this._writeSettings(obj, opts);
-      }
+      this._cacheObj = _.omit(this._cacheObj, keyPath);
```
I don't think we need omit here since it creates a new object, we should be using _.unset https://lodash.com/docs/4.17.5#unset

```
_.unset(this._cacheObj, keyPath)
```

pandafox commented on lib/settings.js:
```diff
@@ -481,14 +314,19 @@ class Settings extends EventEmitter {
    * @param {string} keyPath
    * @param {Function} handler
-   * @returns {Observer}
+   * @returns {number}
```
number?

pandafox commented on lib/settings.js:
```diff
@@ -481,14 +314,19 @@ class Settings extends EventEmitter {
-    return this._watchValueAtKeyPath(keyPath, handler);
+  unwatch(keyPath, handler) {
+    this._watchHandlers = _.reject(f => f.keyPath === keyPath && f.handler === handler);
   }
```
Since we're returning observer, should we also have a function unwatch(oserver)?

votuanphong1983 committed:
Fix Allen's comments:

 - use unset instead of omit
 - unwatch(observer) instead of unwatch(key, handler)

votuanphong1983 committed:
- remove observer.dispose() api

- update related testcases

votuanphong1983 committed:
Remove author and license on each files
